midend/lib/Conversion/LowerDIP/LowerDIPPass.cpp

```diff
@@ -187,142 +187,43 @@ class DIPRotate2DOpLowering : public OpRewritePattern<dip::Rotate2DOp> {
-        loc, arith::CmpFPredicate::OGT, tanBound, tanValAbs);
-
-    // For both rotation strategies, tail processing is handled in second call.
-    rewriter.create<scf::IfOp>(
-        loc, transformCond,
-        [&](OpBuilder &builder, Location loc) {
-          dip::shearTransformController(
-              builder, loc, ctx, lowerBounds1, upperBounds1, steps, strideVal,
-              input, output, sinVec, tanVec, inputRowF32Vec, inputColF32Vec,
-              inputCenterYF32Vec, inputCenterXF32Vec, outputCenterYF32Vec,
-              outputCenterXF32Vec, outputRowLastElemF32, outputColLastElemF32,
-              inputRowLastElemF32, inputColLastElemF32, c0, c0F32, c1F32Vec,
-              vectorTy32, stride, f32);
-
-          dip::shearTransformController(
-              builder, loc, ctx, lowerBounds2, upperBounds2, steps,
-              strideTailVal, input, output, sinVec, tanVec, inputRowF32Vec,
-              inputColF32Vec, inputCenterYF32Vec, inputCenterXF32Vec,
-              outputCenterYF32Vec, outputCenterXF32Vec, outputRowLastElemF32,
-              outputColLastElemF32, inputRowLastElemF32, inputColLastElemF32,
-              c0, c0F32, c1F32Vec, vectorTy32, stride, f32);
-
-          builder.create<scf::YieldOp>(loc);
-        },
-        [&](OpBuilder &builder, Location loc) {
-          dip::standardRotateController(
-              builder, loc, ctx, lowerBounds1, upperBounds1, steps, strideVal,
-              input, output, sinVec, angleVal, inputRowF32Vec, inputColF32Vec,
-              inputCenterYF32Vec, inputCenterXF32Vec, outputCenterYF32Vec,
-              outputCenterXF32Vec, outputRowLastElemF32, outputColLastElemF32,
-              inputRowLastElemF32, inputColLastElemF32, c0, c0F32, c1F32Vec,
-              vectorTy32, stride, f32);
-
-          dip::standardRotateController(
-              builder, loc, ctx, lowerBounds2, upperBounds2, steps,
-              strideTailVal, input, output, sinVec, angleVal, inputRowF32Vec,
-              inputColF32Vec, inputCenterYF32Vec, inputCenterXF32Vec,
-              outputCenterYF32Vec, outputCenterXF32Vec, outputRowLastElemF32,
-              outputColLastElemF32, inputRowLastElemF32, inputColLastElemF32,
-              c0, c0F32, c1F32Vec, vectorTy32, stride, f32);
-
-          builder.create<scf::YieldOp>(loc);
-        });
+    // let alpha = scale * cos(angle), beta = scale * sin(angle)
+    // the affine matrix would be as follow:
+    // [[alpha, beta, (1 - alpha) * centerx - beta * centery],
+    //  [-beta, alpha, beta * centerx + (1 - alpha) * centery]]
+    Value centerX = rewriter.create<arith::ShRSIOp>(loc, inputCol, c1);
+    Value centerY = rewriter.create<arith::ShRSIOp>(loc, inputRow, c1);
+    Value centerXF32 = indexToF32(rewriter, loc, centerX);
+    Value centerYF32 = indexToF32(rewriter, loc, centerY);
+
+    auto affineMatrix = dip::getRotationMatrix(rewriter, loc, centerXF32, centerYF32, angleVal, c1F32);
```

Comment:
It seems only f32 inputs are currently supported. Will we support all types of input images here? If we're going to start with just f32 for now, maybe we should add a TODO comment stating so!